### PR TITLE
dynamically check for conpty support

### PR DESCRIPTION
**Pull request by bagajjal**

### CODE REVIEW

File: contrib/win32/win32compat/win32_pty.c [hunk 1]
```diff
@@ -90,10 +90,39 @@ CreateConPty(const wchar_t *cmdline,
 	return 0;
 }
 
-int is_conpty_supported()
+int
+is_conpty_supported()
 {
-	/* TODO - put conditional logic in here that determines if conpty is supported */
-	return 0;
+	wchar_t system32_path[PATH_MAX] = { 0, };
```
Comment (manojampalam) [inline contrib/win32/win32compat/win32_pty.c:96]: Can you turn this into a singleton? We could just figure it out once whether ConPTY is supported/not.

**Review by manojampalam — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/win32_pty.c (hunk 1 above)
Comment (bagajjal) [reply contrib/win32/win32compat/win32_pty.c:96]: fixed

File: contrib/win32/win32compat/win32_pty.c [hunk 2]
```diff
@@ -90,10 +92,42 @@ CreateConPty(const wchar_t *cmdline,
...
+	if (!GetSystemDirectoryW(system32_path, PATH_MAX)) {
+		error("failed to get system directory");
+		goto done;
+	}
+
+	wcscat_s(kernelbase_dll_path, PATH_MAX, system32_path);
+	wcscat_s(kernelbase_dll_path, PATH_MAX, L"\\KernelBase.dll");
+
+	if ((hm_kernelbase = LoadLibraryW(kernelbase_dll_path)) == NULL) {
+		error("failed to load kernerlbase dll:%s", kernelbase_dll_path);
```
Comment (bingbing8) [inline contrib/win32/win32compat/win32_pty.c:114]: When it is failed to load this dll, does it mean conpty is not supported or it is error? I asked this to find out if an error or debug info should be reported here.

Comment (bagajjal) [reply contrib/win32/win32compat/win32_pty.c:114]: This DLL is supposed to be there from windows7.
Incase if we couldn't find the DLL then we treat as if conpty is not supported.

File: contrib/win32/win32compat/win32_pty.c [hunk 3]
```diff
@@ -98,6 +100,9 @@ is_conpty_supported()
 	HMODULE hm_kernelbase = NULL;
 	int retVal = 0;
 
+	if (isConpty != -1)
+		return isConpty;
```
Comment (manojampalam) [inline contrib/win32/win32compat/win32_pty.c:102]: and where is isConpty being assigned :) ?

**Review by manojampalam — CHANGES_REQUESTED** [review]

**Review by manojampalam — APPROVED** [review]

**Comment by DHowett-MSFT** [comment]
`kernelbase` is not a public windows export library. `CreatePseudoConsole` is exported from `kernel32`. `kernel32` is the public library you should be checking.

`kernelbase` is an implementation detail.

**Comment by bagajjal** [comment]
@DHowett-MSFT - Thanks for leaving a comment. Based on earlier discussion, I used Kernerlbase.dll. I will float the new code review shortly with the fix.

**Comment by DHowett-MSFT** [comment]
Thanks! Sorry for any confusion. 😄

**Comment by bagajjal** [comment]
@DHowett-MSFT - Please review https://github.com/PowerShell/openssh-portable/pull/356.

**Pull request merged**